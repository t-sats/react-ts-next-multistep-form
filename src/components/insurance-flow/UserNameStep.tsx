import React, {useState} from "react";
import { useFormState } from "@/context/FormContext";
import { validateName } from "@/helpers/validation";

const UserNameStep = () => {

  const [error, setError] = useState("");

  const { formData, handleChange, handleNext, handleBack }  = useFormState();

  const firstNameValue = formData.firstName;
  const lastNameValue = formData.lastName;

  const handleOnFocus = () => {
      setError("");
  };

  const handleSubmit = () => {
      if (!validateName(firstNameValue) || !validateName(lastNameValue)) {
          setError("Please enter valid First Name or Last Name");
      } else {
          handleNext();
      }
  };

  return (
    <div className="flex flex-col gap-4 w-full">
      <div className="flex flex-col gap-2 items-start">
        <label htmlFor="firstName">First Name:</label>
        <input
          id="firstName"
          type="text"
          onChange={handleChange}
          onFocus={handleOnFocus}
          name="firstName"
          value={formData.firstName}
          className="input input-primary w-full max-w-xs"
        />
      </div>
      <div className="flex flex-col gap-2 items-start">
        <label htmlFor="lastName">Last Name:</label>
        <input
          id="lastName"
          type="text"
          onChange={handleChange}
          onFocus={handleOnFocus}
          name="lastName"
          value={formData.lastName}
          className="input input-primary w-full max-w-xs"
        />
          {error}
      </div>
      <button
        type="button"
        className="rounded-full border border-solid border-transparent bg-foreground text-background gap-2 hover:bg-[#28a745] font-medium text-sm sm:text-base h-10 sm:h-12 px-4 sm:px-5 sm:w-auto"
        onClick={handleBack}
      >
        Back
      </button>
      <button
        className="rounded-full border border-solid border-transparent bg-foreground text-background gap-2 hover:bg-[#28a745] font-medium text-sm sm:text-base h-10 sm:h-12 px-4 sm:px-5 sm:w-auto"
        type="button"
        onClick={handleSubmit}
      >
        Next
      </button>
    </div>
  );
};

export default UserNameStep;
