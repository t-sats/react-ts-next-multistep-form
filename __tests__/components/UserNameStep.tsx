import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import UserNameStep from "@/components/insurance-flow/UserNameStep";

test('shows an error for incorrect firstName value when the next button is clicked', () => {
  const { getByText, getByLabelText } = render(<UserNameStep />);
  const firstName = getByLabelText("First Name:");
  const nextButton = getByText("Next");
  const error = "Please enter valid First Name or Last Name";

  fireEvent.change(firstName, { target: { value: "test" } });
  fireEvent.click(nextButton);

  expect(getByText(error)).toBeInTheDocument();
});

test('shows an error for empty firstName value when the next button is clicked', () => {
  const { getByText, getByLabelText } = render(<UserNameStep />);
  const firstName = getByLabelText("First Name:");
  const nextButton = getByText("Next");
  const error = "Please enter valid First Name or Last Name";

  fireEvent.change(firstName, { target: { value: "" } });
  fireEvent.click(nextButton);

  expect(getByText(error)).toBeInTheDocument();
});

it('renders UserNameStep component correctly', () => {
  const { container } = render(<UserNameStep />)

  expect(container).toMatchSnapshot()
});
