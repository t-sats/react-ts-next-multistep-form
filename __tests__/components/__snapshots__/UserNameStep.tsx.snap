// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders UserNameStep component correctly 1`] = `
<div>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label
        for="firstName"
      >
        First Name:
      </label>
      <input
        class="input input-primary w-full max-w-xs"
        id="firstName"
        name="firstName"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label
        for="lastName"
      >
        Last Name:
      </label>
      <input
        class="input input-primary w-full max-w-xs"
        id="lastName"
        name="lastName"
        type="text"
        value=""
      />
    </div>
    <button
      class="rounded-full border border-solid border-transparent bg-foreground text-background gap-2 hover:bg-[#28a745] font-medium text-sm sm:text-base h-10 sm:h-12 px-4 sm:px-5 sm:w-auto"
      type="button"
    >
      Back
    </button>
    <button
      class="rounded-full border border-solid border-transparent bg-foreground text-background gap-2 hover:bg-[#28a745] font-medium text-sm sm:text-base h-10 sm:h-12 px-4 sm:px-5 sm:w-auto"
      type="button"
    >
      Next
    </button>
  </div>
</div>
`;
